### ISSUE TITLE
[Bug] use_fused_kernels=True causes training quality regression when ulysses_sequence_parallel_size > 1 for Qwen3-VL GRPO

### ISSUE THREAD

**Issue by IANNXANG** [verl-project/verl#6068]
### System Info

----------Python Info----------
Version      : 3.12.3
Compiler     : GCC 13.3.0
Build        : ('main', 'Nov  3 2025 11:03:51')
Arch         : ('64bit', 'ELF')

------------Pip Info-----------
Version      : 24.0
Directory    : /usr/local/lib/python3.12/site-packages/pip
vllm         : 0.11.0+cu129
sglang       : not found.
ray          : 2.49.2
torch        : 2.8.0

----------verl Info-----------
Version      : 0.8.0.dev
Directory    : /[redacted]/verl/verl
Commit Hash  : 4b9c14f2c306d43f84eccf47315323e3f1ce1270

----------Platform Info----------
Platform     : Linux-5.10.134-013.7.2.[redacted].x86_64-x86_64-with-glibc2.39
system       : Linux
node         : [redacted]
release      : 5.10.134-013.7.2.[redacted].x86_64
version      : #1 SMP Sat Feb 28 03:27:43 UTC 2026

----------Environment----------
CUDA Runtime : 12.9
CUDA Compiler : Cuda compilation tools, release 12.9, V12.9.0

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

We observed this issue while training a vision-language GRPO task in `verl`.

- Model: `Qwen/Qwen3-VL-8B-Instruct`
- Dataset: an internal anonymized multimodal multiple-choice dataset (similar in task format to `Geo3K`)
- Input: text + image
- Output: text
- Reward: based on whether the final multiple-choice answer is correct, using `\boxed{}` extraction / detection

This issue does not manifest as a crash or exception. The training job runs successfully, but the problem is a **consistent training quality regression**: when `ulysses_sequence_parallel_size > 1` is combined with `use_fused_kernels=True`, both reward and validation score keep decreasing.

<img width="225" height="230" alt="Image" src="https://github.com/user-attachments/assets/0854896f-df5d-42f7-9f88-7e00257a7848" />

About the attached figure:

- The rising curves are all `SP=1`
- The descending curves are all `SP>1 + use_fused_kernels=True`, where `SP>1` includes `SP=2`, `4`, and `8`

We also performed step-by-step ablations over many related parameters, including different `batch_size` settings and multiple relevant feature toggles. After these controlled comparisons, the issue consistently pointed to `use_fused_kernels`.

Steps to reproduce:

1. Use the same training setup with:

```bash
train_sp=1
actor_rollout_ref.model.use_fused_kernels=True
actor_rollout_ref.actor.ulysses_sequence_parallel_size=${train_sp}
actor_rollout_ref.ref.ulysses_sequence_parallel_size=${train_sp}
```

Result on our side: training behaves normally, reward increases, and validation score improves.

2. Keep the rest of the training setup unchanged, but change it to:

```bash
train_sp=2   # also reproduced with 4 and 8
actor_rollout_ref.model.use_fused_kernels=True
actor_rollout_ref.actor.ulysses_sequence_parallel_size=${train_sp}
actor_rollout_ref.ref.ulysses_sequence_parallel_size=${train_sp}
```

Result on our side: training still runs successfully, but reward and validation score consistently decrease.

3. Keep the rest of the training setup unchanged, keep `SP>1`, but disable fused kernels:

```bash
train_sp=2   # also verified with 4 and 8
actor_rollout_ref.model.use_fused_kernels=False
actor_rollout_ref.actor.ulysses_sequence_parallel_size=${train_sp}
actor_rollout_ref.ref.ulysses_sequence_parallel_size=${train_sp}
```

Result on our side: the regression disappears and training behavior returns to normal.

The full training script is as follows:

```bash
#!/usr/bin/env bash
set -xeu

# Install local verl source
pip install --no-deps -e /path/to/local/verl

# Training and validation data
train_path="[/path/to/internal_multimodal_train.parquet]"
test_path="[/path/to/internal_multimodal_val.parquet]"

# Base model
MODEL_BASE="Qwen/Qwen3-VL-8B-Instruct"

# Project / experiment name
PROJECT_NAME="ANON_MM_TASK"
EXPR_NAME="$(basename "${MODEL_BASE}")-GRPO_bsz128_TP1SP1_4NODES_0420"

# Output directories
RUN_DIR=/path/to/output_root
DEFAULT_LOCAL_DIR="${RUN_DIR}/checkpoints/${PROJECT_NAME}/${EXPR_NAME}"
ROLLOUT_DATA_DIR="${RUN_DIR}/log/${PROJECT_NAME}/${EXPR_NAME}/rollout"
VALIDATION_DATA_DIR="${RUN_DIR}/log/${PROJECT_NAME}/${EXPR_NAME}/validation"
export TENSORBOARD_DIR="${RUN_DIR}/tensorboard_log/${PROJECT_NAME}/${EXPR_NAME}"

# Generation parameters
max_prompt_length=16384
max_response_length=4096
temperature=1.0
top_p=1.0
top_k=-1 # 0 for HF rollout, -1 for vLLM rollout
val_top_p=0.7

# Algorithm parameters
train_batch_size=128
train_mini_batch_size=32
clip_ratio_low=0.2
clip_ratio_high=0.28
clip_ratio_c=3.0
entropy_coeff=0
grad_clip=1.0
loss_agg_mode=token-mean
use_kl_in_reward=False
kl_coef=0.001
use_kl_loss=False
kl_loss_coef=0.001

# Engineering parameters, mainly related to speed / memory / OOM
rollout_tp=1
train_sp=1
use_dynamic_bsz=True
max_token_len=$((max_prompt_length + max_response_length))
actor_ppo_max_token_len=$(( max_token_len / train_sp ))
infer_ppo_max_token_len=$(( max_token_len / train_sp ))
rollout_max_num_batched_tokens=${max_token_len}
offload=True

python3 -m verl.trainer.main_ppo \
    algorithm.adv_estimator=grpo \
    algorithm.use_kl_in_reward=${use_kl_in_reward} \
    algorithm.kl_ctrl.kl_coef=${kl_coef} \
    data.train_files=${train_path} \
    data.val_files=${test_path} \
    data.train_batch_size=${train_batch_size} \
    data.max_prompt_length=${max_prompt_length} \
    data.max_response_length=${max_response_length} \
    data.filter_overlong_prompts=True \
    data.truncation='error' \
    data.image_key=images \
    actor_rollout_ref.model.path=${MODEL_BASE} \
    actor_rollout_ref.actor.optim.lr=1e-6 \
    actor_rollout_ref.model.use_remove_padding=True \
    actor_rollout_ref.model.use_fused_kernels=False \
    actor_rollout_ref.model.enable_gradient_checkpointing=True \
    actor_rollout_ref.actor.ppo_mini_batch_size=${train_mini_batch_size} \
    actor_rollout_ref.actor.ppo_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.actor.use_dynamic_bsz=${use_dynamic_bsz} \
    actor_rollout_ref.actor.ppo_max_token_len_per_gpu=${actor_ppo_max_token_len} \
    actor_rollout_ref.actor.use_kl_loss=${use_kl_loss} \
    actor_rollout_ref.actor.kl_loss_coef=${kl_loss_coef} \
    actor_rollout_ref.actor.clip_ratio_low=${clip_ratio_low} \
    actor_rollout_ref.actor.clip_ratio_high=${clip_ratio_high} \
    actor_rollout_ref.actor.clip_ratio_c=${clip_ratio_c} \
    actor_rollout_ref.actor.entropy_coeff=${entropy_coeff} \
    actor_rollout_ref.actor.grad_clip=${grad_clip} \
    actor_rollout_ref.actor.loss_agg_mode=${loss_agg_mode} \
    actor_rollout_ref.actor.fsdp_config.param_offload=${offload} \
    actor_rollout_ref.actor.fsdp_config.optimizer_offload=${offload} \
    actor_rollout_ref.rollout.tensor_model_parallel_size=${rollout_tp} \
    actor_rollout_ref.actor.ulysses_sequence_parallel_size=${train_sp} \
    actor_rollout_ref.ref.ulysses_sequence_parallel_size=${train_sp} \
    actor_rollout_ref.rollout.name=vllm \
    +actor_rollout_ref.rollout.engine_kwargs.vllm.disable_mm_preprocessor_cache=True \
    actor_rollout_ref.rollout.gpu_memory_utilization=0.8 \
    actor_rollout_ref.rollout.enable_chunked_prefill=True \
    actor_rollout_ref.rollout.enforce_eager=False \
    actor_rollout_ref.rollout.free_cache_engine=True \
    actor_rollout_ref.rollout.log_prob_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.rollout.log_prob_use_dynamic_bsz=${use_dynamic_bsz} \
    actor_rollout_ref.rollout.log_prob_max_token_len_per_gpu=${infer_ppo_max_token_len} \
    actor_rollout_ref.rollout.max_num_batched_tokens=${rollout_max_num_batched_tokens} \
    actor_rollout_ref.rollout.temperature=${temperature} \
    actor_rollout_ref.rollout.top_p=${top_p} \
    actor_rollout_ref.rollout.top_k=${top_k} \
    actor_rollout_ref.rollout.val_kwargs.temperature=${temperature} \
    actor_rollout_ref.rollout.val_kwargs.top_p=${val_top_p} \
    actor_rollout_ref.rollout.val_kwargs.top_k=${top_k} \
    actor_rollout_ref.rollout.val_kwargs.do_sample=True \
    actor_rollout_ref.rollout.val_kwargs.n=1 \
    actor_rollout_ref.rollout.n=8 \
    actor_rollout_ref.ref.log_prob_micro_batch_size_per_gpu=1 \
    actor_rollout_ref.ref.log_prob_use_dynamic_bsz=${use_dynamic_bsz} \
    actor_rollout_ref.ref.log_prob_max_token_len_per_gpu=${infer_ppo_max_token_len} \
    actor_rollout_ref.ref.fsdp_config.param_offload=${offload} \
    trainer.critic_warmup=0 \
    trainer.logger='["console","tensorboard"]' \
    trainer.project_name=${PROJECT_NAME} \
    trainer.experiment_name=${EXPR_NAME} \
    trainer.default_local_dir=${DEFAULT_LOCAL_DIR} \
    trainer.rollout_data_dir=${ROLLOUT_DATA_DIR} \
    trainer.validation_data_dir=${VALIDATION_DATA_DIR} \
    +ray_kwargs.ray_init.runtime_env.env_vars.TENSORBOARD_DIR=${TENSORBOARD_DIR} \
    trainer.n_gpus_per_node=16 \
    trainer.nnodes=4 \
    trainer.save_freq=20 \
    trainer.test_freq=5 \
    trainer.total_epochs=100 $@

```

There is no error message or stack trace, because this is not a runtime failure but a correctness / quality issue during training.

### Expected behavior

`ulysses_sequence_parallel_size` should mainly be a parallelism / engineering knob, and under the same overall training setup it should not systematically damage training quality.

More specifically, changing from `SP=1` to `SP=2`, `4`, or `8` should not cause reward and validation score to consistently degrade only when `use_fused_kernels=True`; at minimum, the same setup should not exhibit such a stable quality regression.

If the combination of `use_fused_kernels` and `SP>1` is currently unsupported for some VLM paths, it would be helpful to either:

- raise a clear error or warning
- document the limitation explicitly
- or automatically fall back to a safe implementation on the affected path
